### PR TITLE
492 Implement baseline gossiplb algorithm

### LINKED ISSUE #492: Implement GossipLB
**What Needs to be Done?**

- [ ] Finish implementing the GossipLB

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #492.

### CODE REVIEW

File: src/vt/vrt/collection/balance/gossiplb/gossiplb.h [hunk 1]
```diff
@@ -70,18 +76,43 @@ struct GossipLB : BaseLB {
...
-  objgroup::proxy::Proxy<GossipLB> proxy = {};
+  std::unordered_map<NodeType, LoadType> load_info_ = {};
+  objgroup::proxy::Proxy<GossipLB> proxy            = {};
+  bool is_overloaded_                               = false;
+  bool is_underloaded_                              = false;
+  std::unordered_set<NodeType> selected_            = {};
+  std::unordered_set<NodeType> underloaded_         = {};
+  std::unordered_map<ObjIDType, TimeType> cur_objs  = {};
+  LoadType this_new_load                            = 0.0;
+  CriterionEnum criterion_                          = CriterionEnum::ModifiedGrapevine;
```
Comment (nlslatt) [inline src/vt/vrt/collection/balance/gossiplb/gossiplb.h]: Several private data members are missing trailing underscores, which decreases code readability.

File: src/vt/vrt/collection/balance/gossiplb/criterion.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+*/
+
+#if !defined INCLUDED_VT_VRT_COLLECTION_BALANCE_GOSSIPLB_CRITERION_H
+#define INCLUDED_VT_VRT_COLLECTION_BALANCE_GOSSIPLB_CRITERION_H
+
+#include "vt/config.h"
+
+namespace vt { namespace vrt { namespace collection { namespace lb {
+
+enum struct CriterionEnum : int8_t {
```
Comment (pnstickne) [inline src/vt/vrt/collection/balance/gossiplb/criterion.h]: `uint`?

File: src/vt/vrt/collection/balance/gossiplb/criterion.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+struct Criterion : CriterionBase {
+  explicit Criterion(CriterionEnum const criterion)
+    : criterion_(criterion)
+  { }
+
+  bool operator()(LoadType over, LoadType under, LoadType obj, LoadType avg) const {
+    switch (criterion_) {
+    case CriterionEnum::Grapevine:
+      return GrapevineCriterion()(over, under, obj, avg);
+      break;
```
Comment (pnstickne) [inline src/vt/vrt/collection/balance/gossiplb/criterion.h]: `break` after `return`? I've seen this; would be warnings in C# for unreachable-code-detected.

File: src/vt/vrt/collection/balance/gossiplb/gossip_msg.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    s | node_load_;
+  }
+
+private:
+  NodeType from_node_     = uninitialized_destination;
+  NodeLoadType node_load_ = {};
+};
+
+struct LazyMigrationMsg : vt::Message {
+  using ObjsType = std::unordered_map<lb::BaseLB::ObjIDType, lb::BaseLB::LoadType>;
```
Comment (pnstickne) [inline src/vt/vrt/collection/balance/gossiplb/gossip_msg.h]: Is simpler type for message data? If limited range..?

File: src/vt/vrt/collection/balance/gossiplb/gossiplb.h [hunk 5]
```diff
@@ -70,18 +76,43 @@ struct GossipLB : BaseLB {
...
+  void lazyMigrateObjsTo(EpochType epoch, NodeType node, ObjsType const& objs);
+  void inLazyMigrations(balance::LazyMigrationMsg* msg);
+  void thunkMigrations();
 
 private:
-  uint8_t f               = 0;
-  // uint8_t k               = 0;
-  // uint8_t k_cur           = 0;
+  uint8_t f                                         = 2;
+  uint8_t k_max                                     = 4;
```
Comment (pnstickne) [inline src/vt/vrt/collection/balance/gossiplb/gossiplb.h]: `x_`?

File: src/vt/vrt/collection/balance/gossiplb/criterion.h (hunk 3 above)
Comment (lifflander) [reply src/vt/vrt/collection/balance/gossiplb/criterion.h]: Interestingly enough, I believe I get a warning without the `break`, but I will test. Was trying to follow the idiomatic form here (obviously break not needed).

File: src/vt/vrt/collection/balance/gossiplb/gossip_msg.h (hunk 4 above)
Comment (lifflander) [reply src/vt/vrt/collection/balance/gossiplb/gossip_msg.h]: Not sure if I understand the comment. Can you elaborate?

File: src/vt/vrt/collection/balance/gossiplb/criterion.h (hunk 2 above)
Comment (lifflander) [reply src/vt/vrt/collection/balance/gossiplb/criterion.h]: Fixed.

File: src/vt/vrt/collection/balance/gossiplb/gossiplb.h (hunk 1 above)
Comment (lifflander) [reply src/vt/vrt/collection/balance/gossiplb/gossiplb.h]: Good point, I screwed that up. Fixed.

**Comment by nlslatt** [comment]
@ppebay Did you compare the simulator output to the stats files I sent you to make sure that they agree?

File: src/vt/vrt/collection/balance/gossiplb/gossiplb.cc [hunk 6]
```diff
@@ -45,39 +45,423 @@
...
+
+  debug_print(
+    gossiplb, node,
+    "GossipLB::propagateRound: k_max_={}, k_cur_={}, selected.size()={}, fanout={}\n",
+    k_max_, k_cur_, selected.size(), fanout
+  );
+
+  for (int i = 0; i < fanout; i++) {
+    // This implies full knowledge of all processors
+    if (selected.size() >= num_nodes - 1) {
```
Comment (nlslatt) [inline src/vt/vrt/collection/balance/gossiplb/gossiplb.cc]: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]

Comment (lifflander) [reply src/vt/vrt/collection/balance/gossiplb/gossiplb.cc]: Fixed, and pushed back to 1.0.0-beta.4-proposed-update

**Comment by lifflander** [comment]
> @ppebay Did you compare the simulator output to the stats files I sent you to make sure that they agree?

This is a bit difficult because the algorithm uses a random seed. We need to figure out how to cross-validate the resulting partition.

**Comment by lifflander** [comment]
Important: once #561 is merged, we must pull in `776797e7f0d7c5bcc313d97b149bee04100c8df0`

**Comment by lifflander** [comment]
Important: once #561 is merge we must pull in `776797e7f0d7c5bcc313d97b149bee04100c8df0 ` *and* `6f57b719510bc9910d121ef04fc954b05415b049`

**Comment by lifflander** [comment]
> Important: once #561 is merge we must pull in `776797e7f0d7c5bcc313d97b149bee04100c8df0 ` _and_ `6f57b719510bc9910d121ef04fc954b05415b049`

These have been pulled in now.

File: src/vt/vrt/collection/balance/baselb/baselb.cc [hunk 7]
```diff
@@ -228,21 +228,23 @@ void BaseLB::transferMigrations(TransferMsg<TransferVecType>* msg) {
 }
 
 void BaseLB::migrateObjectTo(ObjIDType const obj_id, NodeType const to) {
-  auto& migrator = balance::ProcStats::proc_migrate_;
-  auto iter = migrator.find(obj_id);
   auto from = objGetNode(obj_id);
+  if (from != to) {
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/baselb/baselb.cc]: Stylistically, I have a slight preference for

```
if (from == to)
  return;
```

And then let the rest of the body handle the non-early-exit case.

Many CS professors have hangups about blocks being single-entry, single-exit, but I think that's an obsession for the theoretical benefit of old compilers at the expense of human readability.

**Pull request merged**